src/ring_theory/integral_closure.lean

```diff
@@ -281,3 +282,87 @@ begin
   replace hmp := congr_arg subtype.val hmp,
   exact subtype.eq hmp
 end
+
+end
+
+section algebra
+open algebra
+variables {R : Type*} {A : Type*} {B : Type*}
+variables [comm_ring R] [comm_ring A] [comm_ring B]
+variables [algebra R A] [algebra A B]
+variables (A_alg : ∀ x : A, is_integral R x)
+
+include A_alg
+
+set_option class.instance_max_depth 50
+
+lemma is_integral_trans_aux (x : B) {p : polynomial A} (pmonic : monic p) (hp : aeval A B x p = 0)
+  (S : set (comap R A B))
+  (hS : S = (↑((finset.range (p.nat_degree + 1)).image
+    (λ i, to_comap R A B (p.coeff i))) : set (comap R A B))) :
+  is_integral (adjoin R S) (comap.to_comap R A B x) :=
+begin
+  have coeffs_mem : ∀ i, coeff (map (to_comap R A B) p) i ∈ adjoin R S,
+  { intro i,
+    by_cases hi : i ∈ finset.range (p.nat_degree + 1),
+    { apply algebra.subset_adjoin, subst S,
+      rw [finset.mem_coe, finset.mem_image, coeff_map],
+      exact ⟨i, hi, rfl⟩ },
+    { rw [finset.mem_range, not_lt] at hi,
+      rw [coeff_map, coeff_eq_zero_of_nat_degree_lt hi, alg_hom.map_zero],
+      exact submodule.zero_mem (adjoin R S : submodule R (comap R A B)) } },
+  obtain ⟨q, hq⟩ : ∃ q : polynomial (adjoin R S), q.map (algebra_map (comap R A B)) =
+      (p.map $ to_comap R A B),
```

Comment:
It feels like you should be able to accomplish everything so far using `polynomial.to_subring`.